### PR TITLE
Support link `--overwrite` with `brew "formula", link: :overwrite`

### LINKED ISSUE #1062: feature request: support brew link --overwrite
To handle cases like:
```
Using kubernetes-cli
Linking kubernetes-cli formula.
Linking /usr/local/Cellar/kubernetes-cli/1.23.4...
Error: Could not symlink bin/kubectl
Target /usr/local/bin/kubectl
already exists. You may want to remove it:
  rm '/usr/local/bin/kubectl'

To force the link and overwrite all conflicting files:
  brew link --overwrite kubernetes-cli

To list all files that would be deleted:
  brew link --overwrite --dry-run kubernetes-cli
```

### PR DESCRIPTION
**Pull request by jeremy**
Implies `--force` for uniform usage with keg-only formulae.

Closes #1062.

### CODE REVIEW

**Review by jacobbednarz — APPROVED** [review]
this seems reasonable to me given the previous discussion on it. @MikeMcQuaid do you have any reservations about the API here?

File: lib/bundle/brew_installer.rb [hunk 1]
```diff
@@ -92,6 +92,9 @@ def service_change_state!(verbose:)
 
     def link_change_state!(verbose: false)
       case @link
+      when :overwrite
+        puts "Overwriting links for #{@name} formula." if verbose
+        Bundle.system(HOMEBREW_BREW_FILE, "link", "--force", "--overwrite", @name, verbose: verbose)
```
Comment (MikeMcQuaid) [inline lib/bundle/brew_installer.rb:97]: Are you sure `--force` is needed here? That should only be needed if it's keg-only.

**Review by MikeMcQuaid — COMMENTED** [review]
I'm 👍🏻 with this being done. I do wonder if we want a similar DSL/status for casks to be consistent there?

File: lib/bundle/brew_installer.rb (hunk 1 above)
Comment (jeremy) [reply lib/bundle/brew_installer.rb:97]: Right - it's meant to support overwriting links with a keg-only formula. Given that `--force` is effectively a no-op for non-key-only formulae, I figured it's clearer to use a single invocation that works for any formula.

Comment (MikeMcQuaid) [reply lib/bundle/brew_installer.rb:97]: @jeremy I think `--force` should only be passed when it's keg-only, we have that information already in memory and it feels like a cleaner implementation.

Comment (jeremy) [reply lib/bundle/brew_installer.rb:97]: That state ("keg-only and unlinked") is not presently available, but could be with changes to BrewDumper.

BrewDumper.formulae sets `link?` to one of three values which capture four distinct states:
| dumped | logic | keg_only | linked |
|---|---|---|---|
| `link?: true` | keg_only? && linked? | true | true |
| `link?: false` | !keg_only? && !linked? | false | false |
| `link?: nil` | keg_only? && !linked? | true | false |
| `link?: nil` | !keg_only? && linked? | false | true |

We could have BrewDumper use `link?: :overwrite` to capture the `keg_only? && !linked?` state, or we could split the compound `link?` state into `linked?` and `keg_only?` booleans.

Comment (MikeMcQuaid) [reply lib/bundle/brew_installer.rb:97]: > We could have BrewDumper use link?: :overwrite to capture the keg_only? && !linked? state, or we could split the compound link? state into linked? and keg_only? booleans.

Fine with either option 👍🏻

**Comment by swrobel** [comment]
Would be sweet to get this merged. It's the one thing holding our team back from using a Brewfile

**Comment by MikeMcQuaid** [comment]
@swrobel Note sure why the 😕 response here. You commented on a closed PR where I've given feedback on what needs to be changed. This is open source: we're all volunteers here and if this is blocking your team: you or your team can do the remaining work to get this over the line/

**Comment by swrobel** [comment]
Fair enough. What I saw last night was that it had been approved by @jacobbednarz, yet left unmerged. I didn't realize changes were requested.

**Pull request closed without merging**